### PR TITLE
DHCPv6 improvements

### PR DESCRIPTION
**Pull request by abbradar**
###### Motivation for this change

1. Allow one to simultaneously have two instances of `dhcpd` running for DHCPv4 and DHCPv6 respectively, with separate configuration;
2. (controversial, needs review from someone who understands risks of IPv6 better!) Allow DHCPv6 client incoming traffic in the firewall by default (without this DHCPv6 client won't work);
3. Test radvd+dhcpd6 as part of our networking tests. Client discovers a router, gets an address from the pool and pings itself and the server;
4. Update radvd to the latest version.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/networking/firewall.nix [hunk 1]
```diff
@@ -172,13 +172,16 @@ let
...
     ${optionalString config.networking.enableIPv6 ''
+      # Accept all ICMPv6 messages except redirects and node
+      # information queries (type 139).  See RFC 4890, section
+      # 4.4.
       ip6tables -A nixos-fw -p icmpv6 --icmpv6-type redirect -j DROP
       ip6tables -A nixos-fw -p icmpv6 --icmpv6-type 139 -j DROP
       ip6tables -A nixos-fw -p icmpv6 -j nixos-fw-accept
+
+      # Allow this host to act as a DHCPv6 client
+      ip6tables -A nixos-fw -d fe80::/64 -p udp --dport 546 -j nixos-fw-accept
```
Comment (Mic92) [inline nixos/modules/services/networking/firewall.nix]: I don't think this needs to be enabled on every machine as only the a minority of installations provide `dhcpv6-server`. The option itself looks good to me. I would add it to `extraCommands` in the dhcpv6 service itself.

File: nixos/modules/services/networking/dhcpd.nix [hunk 2]
```diff
@@ -29,131 +28,180 @@ let
...
+          ethernetAddress = "00:16:76:9a:32:1d";
+          ipAddress = "192.168.1.10";
+        }
+        { hostName = "bar";
+          ethernetAddress = "00:19:d1:1d:c4:9a";
+          ipAddress = "192.168.1.11";
+        }
+      ];
+      description = ''
+	A list mapping Ethernet addresses to IPv${postfix} addresses for the
```
Comment (fpletz) [inline nixos/modules/services/networking/dhcpd.nix]: Indention. ;)

**Review by fpletz — APPROVED** [review]
I'm not using the ISC dhcpd much these days and didn't test this myself but the code looks good and there is a test. And more support for IPv6 is always nice. Awesome! :+1:

**Comment by abbradar** [comment]
@fpletz Do you think it's okay to allow DHCPv4 client traffic in firewall unconditionally? I don't see any immediate security problems (it's also enabled in Red Hat by default it seems, not sure about others) but I'm a newcomer to IPv6...

**Comment by fpletz** [comment]
@abbradar I agree that this shouldn't be a security issue. The rule strict enough so only packets from hosts in the same broadcast domain will be accepted (link-local addresses). And the allowed destination port is reserved for `dhcpv6-client` only and is restriced to UDP. It's the job of the DHCPv6 client to listen to the correct interface or check for plausible responses.

**Pull request merged**